File: internal/util/files.ts
import path from "node:path";
import { stat, readdir } from "node:fs/promises";

export interface WalkOptions {
  maxFiles: number;
  maxDepth: number;
  skipDirNames: Record<string, boolean>;
}

export function defaultWalkOptions(): WalkOptions {
  return {
    maxFiles: 5000,
    maxDepth: 30,
    skipDirNames: {
      ".git": true,
      "node_modules": true,
      "vendor": true,
      "dist": true,
      "build": true,
      "bin": true,
      ".agent-harness": true,
      ".agent-harness-cache": true,
    },
  };
}

async function walkDir(
  root: string,
  rel: string,
  opts: WalkOptions,
  out: string[],
): Promise<void> {
  const full = path.join(root, rel);
  const entries = await readdir(full, { withFileTypes: true });
  for (const entry of entries) {
    if (out.length >= opts.maxFiles) return;
    const nextRel = path.join(rel, entry.name);
    const depth = nextRel === "" ? 0 : nextRel.split(path.sep).length;
    if (depth > opts.maxDepth) {
      if (entry.isDirectory()) continue;
    }
    if (entry.isDirectory()) {
      if (opts.skipDirNames[entry.name]) continue;
      await walkDir(root, nextRel, opts, out);
      continue;
    }
    if (!entry.isFile()) continue;
    out.push(path.posix.join(...nextRel.split(path.sep)));
  }
}

export async function walkFiles(root: string, opts: WalkOptions): Promise<string[]> {
  if (!opts.maxFiles || opts.maxFiles <= 0) {
    throw new Error("maxFiles must be > 0");
  }
  if (!opts.maxDepth || opts.maxDepth <= 0) {
    opts.maxDepth = 30;
  }
  await stat(root);
  const out: string[] = [];
  await walkDir(path.resolve(root), "", opts, out);
  return out;
}

File: internal/util/files.py
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List


@dataclass
class WalkOptions:
    max_files: int = 5000
    max_depth: int = 30
    skip_dir_names: Dict[str, bool] = field(default_factory=dict)


def default_walk_options() -> WalkOptions:
    return WalkOptions(
        max_files=5000,
        max_depth=30,
        skip_dir_names={
            ".git": True,
            "node_modules": True,
            "vendor": True,
            "dist": True,
            "build": True,
            "bin": True,
            ".agent-harness": True,
            ".agent-harness-cache": True,
        },
    )


def walk_files(root: str, opts: WalkOptions) -> List[str]:
    if not opts.max_files or opts.max_files <= 0:
        raise ValueError("max_files must be > 0")
    if not opts.max_depth or opts.max_depth <= 0:
        opts.max_depth = 30

    base = Path(root).resolve()
    if not base.exists():
        raise FileNotFoundError(root)

    out: List[str] = []

    def walk_dir(current: Path, rel: Path) -> None:
        if len(out) >= opts.max_files:
            return
        try:
            entries = list(current.iterdir())
        except OSError:
            return
        for entry in entries:
            if len(out) >= opts.max_files:
                return
            next_rel = rel / entry.name
            depth = len(next_rel.parts)
            if depth > opts.max_depth and entry.is_dir():
                continue
            if entry.is_dir():
                if opts.skip_dir_names.get(entry.name):
                    continue
                walk_dir(entry, next_rel)
                continue
            if not entry.is_file():
                continue
            out.append(next_rel.as_posix())

    walk_dir(base, Path(""))
    return out
